src/lib/salad/hll.c

```diff
@@ -0,0 +1,1307 @@
+ * determined by the algorithm precision parameter.
+ * For the dense representation the relative error is 1.04/sqrt(m) and the
+ * memory capacity is m*6 bits where m is number of counters wich
+ * equals to 2^precision.
+ * For the sparse representation the memory usage is proportional to the number
+ * of distinct elements that has added until it reaches the memory usage of the
+ * dense representation, and then it switches to the dense representation with
+ * fixed memory requretments. The sparse representation has the best accuracy.
+ */
+struct hll {
+	/** See the comment to HLL_REPRESENTATION enum. */
+	enum HLL_REPRESENTATION representation;
+	/**
+	 * Interpretation of the data depends on the representation.
+	 * For dense representation it's an array of registers of size
+	 * 2^precision * 6 bits. Registers store the maximum added rank of set
+	 * of hashes wich last precision bits are equal to register index.
+	 * For the sparse representation it's a sparsely represented
+	 * HyperLogLog. Instead of registers the sparse representation works
+	 * with pairs consisting of the rank and the index of the added hashes.
+	 */
+	uint8_t *data;
+	/**
+	 * Precision is equal to the number of bits that are interpreted as
+	 * a register index. Available values are from HLL_MIN_PRECISION to
+	 * HLL_MAX_PRECISION (defined in hll_emprirical.h.)
+	 * The larger value leads to less estimation error but larger
+	 * memory requirement (2^precision * 6 bits).
+	 * This value is only used by the dense representation and for switching
+	 * from the sparse representation.
+	 */
+	uint8_t dprecision;
+	/**
+	 * Cached value of the last estimation.
+	 */
+	double cached_estimation;
+};
+
+/* Check if the precision is correct. */
+static int MAYBE_UNUSED
+hll_is_valid_precision(uint8_t prec)
+{
+	return hll_is_supported_precision(prec) || prec == HLL_SPARSE_PRECISION;
+}
+
+/* Get a number whose first n bits are equal to ones. */
+static uint64_t
+hll_ones(uint8_t n)
+{
+	assert(n <= 64);
+	if (likely(n < 64))
+		return (UINT64_C(1) << n) - 1;
+	return UINT64_C(-1);
+}
+
+/* Check whether the cached value stores the valid value of the estimation. */
+static int
+hll_is_valid_cache(const struct hll *hll)
+{
+	return hll->cached_estimation >= 0;
+}
+
+/* Mark the cached value as invalid. */
+static void
+hll_invalidate_cache(struct hll *hll)
+{
+	hll->cached_estimation = -1.0;
+}
+
+/*
+ * The highest precision bits of the hash are interpreted as a register index.
+ */
+static uint32_t
+hll_hash_register_idx(uint64_t hash, uint8_t precision)
+{
+	assert(hll_is_valid_precision(precision));
+	return hash >> (64 - precision);
+}
+
+/*
+ * Return the number of trailing zeros bits in hash plus one.
+ * Register bits are interpreted as ones.
+ */
+static uint8_t
+hll_hash_rank(uint64_t hash, uint8_t precision)
+{
+	assert(hll_is_valid_precision(precision));
+	hash |= hll_ones(precision) << (64 - precision);
```

Comment:
You can use `1ull` instead of `hll_ones(precision)`.

Reply:
The implementation from the article assumes that the register bits are not used to calculate the rank.